mqtt/mqtt-bridge/src/persist/storage/mod.rs

```diff
@@ -0,0 +1,136 @@
+use async_trait::async_trait;
+use std::{
+    collections::VecDeque,
+    future::Future,
+    pin::Pin,
+    sync::atomic::{AtomicUsize, Ordering},
+    task::Waker,
+};
+
+pub mod ring_buffer;
+// pub mod sled;
+#[macro_use]
+mod error_macro;
+pub mod error;
+mod serialize;
+
+use error::StorageError;
+use mqtt3::proto::Publication;
+use serialize::binary_deserialize;
+
+use self::serialize::binary_serialize;
+
+use super::{Key, PersistError, StreamWakeableState};
+
+pub type StorageResult<T> = Result<T, StorageError>;
+
+#[derive(Clone, Copy, Debug)]
+pub enum FlushOptions {
+    AfterEachWrite,
+    AfterXWrites(usize),
+    AfterXBytes(usize),
+    AfterXMilliseconds(usize),
+    Off,
+}
+
+pub struct FlushState {
+    writes: AtomicUsize,
+    bytes_written: AtomicUsize,
+    millis_elapsed: AtomicUsize,
+}
+
+impl FlushState {
+    fn new() -> Self {
+        Self {
+            writes: AtomicUsize::default(),
+            bytes_written: AtomicUsize::default(),
+            millis_elapsed: AtomicUsize::default(),
+        }
+    }
+
+    fn reset(&self, flush_option: &FlushOptions) {
+        match flush_option {
+            FlushOptions::AfterEachWrite => {}
+            FlushOptions::AfterXWrites(_) => {
+                self.writes.store(0, Ordering::SeqCst);
+            }
+            FlushOptions::AfterXBytes(_) => {
+                self.bytes_written.store(0, Ordering::SeqCst);
+            }
+            FlushOptions::AfterXMilliseconds(_) => {
+                self.millis_elapsed.store(0, Ordering::SeqCst);
+            }
+            FlushOptions::Off => {}
+        }
+    }
+
+    fn update(&self, writes: usize, bytes_written: usize, millis_elapsed: usize) {
+        self.bytes_written
+            .fetch_add(bytes_written, Ordering::SeqCst);
+        self.millis_elapsed
+            .fetch_add(millis_elapsed, Ordering::SeqCst);
+        self.writes.fetch_add(writes, Ordering::SeqCst);
+    }
+}
+
+pub trait Storage<Key, Value>: Send + Sync {
+    fn insert_future<'a>(
+        &self,
+        value: &Vec<u8>,
+    ) -> Pin<Box<dyn Future<Output = StorageResult<()>> + Send + 'a>>;
+    fn batch_future<'a>(
+        &self,
+        amount: usize,
+    ) -> Pin<Box<dyn Future<Output = StorageResult<VecDeque<(Key, Value)>>> + Send + 'a>>;
+    fn remove_future<'a>(
+        &self,
+        key: &Key,
+    ) -> Pin<Box<dyn Future<Output = StorageResult<()>> + Send + 'a>>;
+    fn set_waker(&mut self, waker: Waker);
+    fn waker(&mut self) -> &mut Option<Waker>;
+}
+
```

Comment:
This interface can likely go away, it is a useful interface for my development purposes if there is some minor changes I want to test I don't have to break the entire code base, just patch it up below in StreamWakeableState.

Reply:
Yes, this seems unnecessary trait at the end. But just in case, in my previous comment I mentioned that we should probably try to stay away from async traits and try not to use dynamic dispatching